### DISCUSSION TITLE
Notes on deploying Marlowe Runtime

### DISCUSSION THREAD

**Discussion by bwbush** [marlowe-lang/marlowe-cardano#293] [DevOps]
This discussion is for notes and discussion related to deploying Marlowe Runtime. This will be a source for creating documentation for deploying Marlowe Runtime.

**Comment by bwbush** [comment]
# Chain Seek

Running the Marlowe chain seek daemon.
1. Install PostgreSQL 13.
2. `createdb` followed by name of the database to use.
3. `nix develop` in top folder of repository.
4. `cd marlowe-sync`.
5. Edit `sqitch.conf` so that it has the username and database to use.
6. `sqitch deploy`.
7. cabal run exe:chainseekd -- --socket-path ../../node.socket --database-uri postgresql://myusername@0.0.0.0/preview_chainseek --genesis-config-file ../../cardano-node/configuration/cardano/mainnet-byron-genesis.json --genesis-config-file-hash 5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb`

**Reply by bwbush** [reply]
Running on `mainnet` requires adjustment to the write-ahead log size in Postgres. I'll post a diff for `postgresql.conf`.

**Comment by bwbush** [comment]
Here are the PostgreSQL parameter changes (relative to default) that I'm using for `mainnet`:
```diff
122c122
< shared_buffers = 128MB                        # min 128kB
---
> shared_buffers = 2GB                  # min 128kB
124c124
< #huge_pages = try                     # on, off, or try
---
> huge_pages = try                      # on, off, or try
126,127c126,127
< #temp_buffers = 8MB                   # min 800kB
< #max_prepared_transactions = 0                # zero disables the feature
---
> temp_buffers = 2GB                    # min 800kB
> max_prepared_transactions = 256               # zero disables the feature
229c229
< max_wal_size = 1GB
---
> max_wal_size = 4GB
723c723
< #max_pred_locks_per_transaction = 64  # min 10
---
> max_pred_locks_per_transaction = 256  # min 10
```

**Comment by bwbush** [comment]
Here is my configuration for `preview`:

```console
$ cat configuration.env 
export MARLOWE_ROOT="$(dirname -- "$(readlink -f -- "${BASH_SOURCE[0]}")")"

export CARDANO_NODE_SOCKET_PATH="$MARLOWE_ROOT/node.socket"
export CARDANO_TESTNET_MAGIC=2

MAGIC=(--testnet-magic 2)

CARDANO_NODE_PORT=23001

CHAINSEEKD_PORT=23715
CHAINSEEKD_QUERY_PORT=23716
CHAINSEEKD_JOB_PORT=23720

MARLOWE_HISTORY_COMMAND_PORT=23717
MARLOWE_HISTORY_QUERY_PORT=23718
MARLOWE_HISTORY_SYNC_PORT=23719

MARLOWE_TX_COMMAND_PORT=23723

MARLOWE_DISCOVERY_QUERY_PORT=23721
MARLOWE_DISCOVERY_SYNC_PORT=23722
```

Set variables:
```bash
. configuration.env
```

Run node:
```bash
build-node/bin/cardano-node run --config config/node.json --topology config/topology.json --database-path node.db --socket-path node.socket --port $CARDANO_NODE_PORT
```

Set chain seek database:
```console
$ createdb mychainseek

$ vi sqitch.conf

$ sqitch deploy
```

Run chain seek:
```bash
chainseekd --genesis-config-file-hash 72593f260b66f26bef4fc50b38a8f24d3d3633ad2e854eaf73039eb9402706f1 --genesis-config-file ../../config/byron-genesis.json --socket-path ../../node.socket --database-uri postgresql://localhost/preview_chainseek ${MAGIC[@]} --port-number $CHAINSEEKD_PORT --query-port-number $CHAINSEEKD_QUERY_PORT --job-port-number $CHAINSEEKD_JOB_PORT
```

Run contract history:
```bash
marlowe-history --chain-seek-port-number $CHAINSEEKD_PORT --chain-seek-query-port-number $CHAINSEEKD_QUERY_PORT --command-port $MARLOWE_HISTORY_COMMAND_PORT --query-port $MARLOWE_HISTORY_QUERY_PORT --sync-port $MARLOWE_HISTORY_SYNC_PORT
```

Run transaction creation:
```bash
marlowe-tx --chain-seek-port-number $CHAINSEEKD_PORT --chain-seek-query-port-number $CHAINSEEKD_QUERY_PORT --chain-seek-command-port-number $CHAINSEEKD_JOB_PORT --command-port $MARLOWE_TX_COMMAND_PORT --history-sync-port $MARLOWE_HISTORY_SYNC_PORT
```

Run contract discovery:
```bash
marlowe-discovery --chain-seek-port-number $CHAINSEEKD_PORT --chain-seek-query-port-number $CHAINSEEKD_QUERY_PORT --query-port $MARLOWE_DISCOVERY_QUERY_PORT --sync-port $MARLOWE_DISCOVERY_SYNC_PORT
```

Use Marlowe Runtime CLI:
```bash
marlowe --history-command-port $MARLOWE_HISTORY_COMMAND_PORT --history-query-port $MARLOWE_HISTORY_QUERY_PORT --history-sync-port $MARLOWE_HISTORY_SYNC_PORT --tx-command-port $MARLOWE_TX_COMMAND_PORT ls
```
```

**Reply by bwbush** [reply]
All of the `MARLOWE_` above should be `MARLOWE_RT_` instead, for compatibility with the default environment variables used by the `marlowe` command.

**Comment by bwbush** [comment]
[Here is a tutorial](https://github.com/input-output-hk/marlowe-cardano/blob/bwbush/mainnet/marlowe-cli/cookbook/runtime-actus-pam.ipynb) for using Marlowe Runtime history, following, and discovery.

**Comment by palas** [comment]
Here is a diagram with how the pieces fit together and default ports:

![Port diagram](https://user-images.githubusercontent.com/638102/195166785-2f227174-8fa7-4c66-8206-d8e3cb5f1d9e.svg)
<details> 
<summary><code>ports.dot</code></summary>
<pre><code>digraph {
&nbsp;  rankdir=LR
&nbsp;  ranksep="2.0"
&nbsp;  ordering=out
&nbsp;  node [shape=record]; 
&nbsp;  subgraph cluster_cardano_node {
&nbsp;    label = "cardano-node";
&nbsp;    subgraph cluster_cardano_node_inp {
&nbsp;      label = "Server" ; 
&nbsp;      cn_unix_socket [ label= "cardano_node: UNIX socket" ];
&nbsp;    }
&nbsp;    subgraph cluster_cardano_node_out {
&nbsp;      label = "Client";
&nbsp;    }
&nbsp;  }
&nbsp;
&nbsp;  subgraph cluster_chain_seek {
&nbsp;    label = "chain-sync"
&nbsp;    subgraph cluster_chain_seek_inp {
&nbsp;      label = "Server"; 
&nbsp;      cs_port_number [ label= "port-number: 3715" ];
&nbsp;      cs_query_port_number [ label= "query-port-number: 3716" ];
&nbsp;      cs_job_port_number [ label= "job-port-number: 3720" ];
&nbsp;    }
&nbsp;    subgraph cluster_chain_seek_out {
&nbsp;      label = "Client";
&nbsp;      cs_cn_unix_socket [ label= "cardano_node: UNIX socket"];
&nbsp;    }
&nbsp;   }
&nbsp;
&nbsp;   subgraph cluster_marlowe_history {
&nbsp;    label = "marlowe-history";
&nbsp;    subgraph cluster_marlowe_history_inp {
&nbsp;      label = "Server" ; 
&nbsp;      mh_command_port [ label= "command-port: 3717" ];
&nbsp;      mh_query_port [ label= "query-port: 3718" ];
&nbsp;      mh_sync_port [ label= "sync-port: 3719" ];
&nbsp;    }
&nbsp;    subgraph cluster_marlowe_history_out {
&nbsp;      label = "Client";
&nbsp;      mh_cs_port_number [ label= "chain-seek-port-number: 3715" ];
&nbsp;      mh_cs_query_server [ label= "chain-seek-query-port-number: 3716" ];
&nbsp;    }
&nbsp;   }
&nbsp;
&nbsp;   subgraph cluster_marlowe_tx {
&nbsp;    label = "marlowe-tx";
&nbsp;    subgraph cluster_marlowe_tx_inp {
&nbsp;      label = "Server" ; 
&nbsp;      mt_command_port [ label= "command-port: 3723" ];
&nbsp;    }
&nbsp;    subgraph cluster_marlowe_tx_out {
&nbsp;      label = "Client";
&nbsp;      mt_cs_port_number [ label= "chain-seek-port-number: 3715" ];
&nbsp;      mt_cs_query_server [ label= "chain-seek-query-port-number: 3716" ];
&nbsp;      mt_cs_command_port [ label= "chain-seek-command-port-number: 3720" ];
&nbsp;      mt_mh_history_sync_port [ label= "history-sync-port: 3719" ];
&nbsp;    }
&nbsp;   }
&nbsp;
&nbsp;   subgraph cluster_marlowe_discovery {
&nbsp;    label = "marlowe-discovery";
&nbsp;    subgraph cluster_marlowe_discovery_inp {
&nbsp;      label = "Server" ; 
&nbsp;      md_query_port [ label= "query-port: 3721" ];
&nbsp;      md_sync_server [ label= "sync-port: 3722" ];
&nbsp;    }
&nbsp;    subgraph cluster_marlowe_discovery_out {
&nbsp;      label = "Client";
&nbsp;      md_cs_port_number [ label= "chain-seek-port-number: 3715" ];
&nbsp;      md_cs_query_server [ label= "chain-seek-query-port-number: 3716" ];
&nbsp;    }
&nbsp;   }
&nbsp;
&nbsp;   subgraph cluster_marlowe {
&nbsp;    label = "marlowe (runtime cli)";
&nbsp;    subgraph cluster_marlowe_inp {
&nbsp;      label = "Server" ; 
&nbsp;    }
&nbsp;    subgraph cluster_marlowe_out {
&nbsp;      label = "Client";
&nbsp;      m_history_command [ label= "history-command: 3717" ];
&nbsp;      m_history_query [ label= "history-query: 3718" ];
&nbsp;      m_history_sync [ label= "history-sync: 3719" ];
&nbsp;      m_discovery_query [ label= "discovery-query: 3721" ];
&nbsp;      m_tx_command [ label= "tx-command: 3723" ];
&nbsp;    }
&nbsp;   }
&nbsp;
&nbsp;  cs_cn_unix_socket -> cn_unix_socket;
&nbsp;  mh_cs_port_number -> cs_port_number;
&nbsp;  mh_cs_query_server -> cs_query_port_number;
&nbsp;  mt_cs_port_number -> cs_port_number;
&nbsp;  mt_cs_query_server -> cs_query_port_number;
&nbsp;  mt_cs_command_port -> cs_job_port_number;
&nbsp;  mt_mh_history_sync_port -> mh_sync_port;
&nbsp;  md_cs_port_number -> cs_port_number;
&nbsp;  md_cs_query_server -> cs_query_port_number;
&nbsp;  m_history_command -> mh_command_port;
&nbsp;  m_history_query -> mh_query_port;
&nbsp;  m_history_sync -> mh_sync_port;
&nbsp;  m_discovery_query -> md_query_port;
&nbsp;  m_tx_command -> mt_command_port;
}
</code></pre>
</details>

**Comment by bwbush** [comment]
Here are comprehensive instructions on setting up the EC2 instance: https://github.com/bwbush/local-cardano-service-configs/blob/marlowe-runtime-ec2/ReadMe.md.

**Comment by bwbush** [comment]
Here are a couple of important notes for workarounds needed in Marlowe Runtime deployments:
* _Marlowe Runtime does not survive epoch boundaries:_ My current hypothesis is that the root cause of SCP-4586 is that Runtime's sense of time gets scrambled after an epoch boundary and transaction-creation will fail with peculiar Ouroboros errors. The workaround is to restart all services after an epoch boundary.
* _Chain seek crashes due to some rollbacks:_ Especially after catching up with the tip after being restarted, chainseekd will crash, SCP-4530. The workaround is to restart chainseekd after it crashes and hope it doesn't crash again.
* _Dependencies on chain seek:_ It appears that marlowe-history, marlowe-discovery, and marlowe-tx can be restarted in any order, but if chainseekd is restarted, then the downstream services may crash or failure to correctly service requests. The workaround is to restart marlowe-history, marlowe-discovery, and marlowe-tx after chainseekd starts listening on its ports after it has been started.
